### ISSUE TITLE
[RFC] Opset Version Support

### ISSUE THREAD

**Issue by tjingrant** [onnx/onnx-mlir#42]
To summarize our discussion on our support for ONNX opset versions,  I'm opening up an RFC thread to discuss and track the status of implementation.

## Motivation:

ONNX is a standard consisting of the semantic definitions of operations commonly used in ML/AI context. Since the standard (and the underlying field) is rapidly evolving, some definitions go out of date, and to distinguish between new/old definitions, a unique version is given to every definition (including the schema) of an operation.

However, the current state of our infrastructure only supports the latest of many definitions, meaning our support for the ONNX standard is partial. The reason leading up to our partial support is that implementing/maintaining operation lowering support entails expensive endeavor. 

Therefore, a proposal to expand our operation lowering support to cover the entirety of ONNX standard must reconcile with the fact that we cannot provide lowering support for all versions of operation semantics/schemas.

## Proposed Solution:

- We continue to support the lowering of only the latest version of every operation, we will continue to name these operations without any version prefix/suffix (e.g., `onnx.Add`).
- We support the representation (thus schema) of all versions of operation, and the older versions of operations will be named with a "-{version}" suffix (e.g., `onnx.Add-6`). 
- For older versions of operations, we convert them, upon encountering, to their corresponding definition/schema of newer versions by invoking pattern-based rewrite rules.

The naming of versioned operations is consistent with that appearing in the ONNX standard (c.f. https://github.com/onnx/onnx/blob/master/docs/Changelog.md#Add-6)

## Caveat:

One caveat of such a naming scheme is that existing IR files may go out of date as new versions of operation are introduced. This problem can be mitigated by attaching the current latest onnx opset version as an attribute to modules containing ONNX dialect.

**Comment by chentong319** [comment]
I agree on the main idea: convert versioned operations to one set of operation with rewrite. Some comments in design:
-  I'd like to have a version-less set of operations in onnx-mlir, instead of the latest version. The onnx-mlir project needs an anchor for transformation/optimization, while the onnx standard may evolve. In reality, the current version-less set happen to be the highest version in CURRENT ONNX standard, and may have some operations which are not directly in onnx. This may be related to the topic of designing the kernel operation set for onnx.
- In implementing, the rewrite of the current version to version-less set is straightforward. We just need to record the current version number for all operations in the rewrite rules in onnx-mlir. 
- For rewriting of other versions, I will investigate whether the version-less set is expressive enough or new operations have to be introduce, and generate rewriting rules.
- The version in onnx is quite sparse: most of operations remain the same for lots of versions. Need to add a version match function when a model is imported. 

 

1. 

 the current version of each operation should be recorded separately inside onnx-mlir.  The Therefore, any newer or older version can be detected.
2.

**Comment by AlexandreEichenberger** [comment]
I like the proposal.  From an IBM customer perspective, we will be mostly interested in backward compatibility back to the version of the first official product. I understand that other companies with products providing compatibility to earlier version may be interested in compatibility to earlier version as well, so setting up a general mechanism that enables compatibility to arbitrary version is very useful.

**Comment by tungld** [comment]
I like the idea of developing a version-less set of operations in onnx-mlir to which versioned operations will be rewritten. The version-less set is expected to be stable and optimization/transformation oriented.
